# 1. 2023년 pdb 가져오기

Link: [RCSB][rcsblink]

[rcsblink]: https://www.rcsb.org/search?request=%7B%22query%22%3A%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22group%22%2C%22logical_operator%22%3A%22and%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22parameters%22%3A%7B%22operator%22%3A%22exact_match%22%2C%22attribute%22%3A%22rcsb_entry_info.experimental_method%22%2C%22value%22%3A%22X-ray%22%7D%2C%22service%22%3A%22text%22%7D%2C%7B%22type%22%3A%22terminal%22%2C%22parameters%22%3A%7B%22operator%22%3A%22range%22%2C%22attribute%22%3A%22rcsb_accession_info.initial_release_date%22%2C%22value%22%3A%7B%22from%22%3A%222023-01-01T00%3A00%3A00Z%22%2C%22to%22%3A%222023-12-31T23%3A59%3A59Z%22%2C%22include_lower%22%3Atrue%2C%22include_upper%22%3Atrue%7D%7D%2C%22service%22%3A%22text%22%7D%5D%7D%2C%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22rcsb_entity_source_organism.ncbi_scientific_name%22%2C%22value%22%3A%22Homo%20sapiens%22%2C%22operator%22%3A%22exact_match%22%7D%7D%5D%2C%22logical_operator%22%3A%22or%22%2C%22label%22%3A%22rcsb_entity_source_organism.ncbi_scientific_name%22%7D%2C%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22exptl.method%22%2C%22value%22%3A%22X-RAY%20DIFFRACTION%22%2C%22operator%22%3A%22exact_match%22%7D%7D%5D%2C%22logical_operator%22%3A%22or%22%2C%22label%22%3A%22exptl.method%22%7D%2C%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22entity_poly.rcsb_entity_polymer_type%22%2C%22value%22%3A%22Protein%22%2C%22operator%22%3A%22exact_match%22%7D%7D%5D%2C%22logical_operator%22%3A%22or%22%2C%22label%22%3A%22entity_poly.rcsb_entity_polymer_type%22%7D%2C%7B%22type%22%3A%22group%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22rcsb_entry_info.resolution_combined%22%2C%22value%22%3A%7B%22from%22%3A0.5%2C%22to%22%3A1%2C%22include_lower%22%3Atrue%2C%22include_upper%22%3Afalse%7D%2C%22operator%22%3A%22range%22%7D%7D%2C%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22rcsb_entry_info.resolution_combined%22%2C%22value%22%3A%7B%22from%22%3A1%2C%22to%22%3A1.5%2C%22include_lower%22%3Atrue%2C%22include_upper%22%3Afalse%7D%2C%22operator%22%3A%22range%22%7D%7D%2C%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22attribute%22%3A%22rcsb_entry_info.resolution_combined%22%2C%22value%22%3A%7B%22from%22%3A1.5%2C%22to%22%3A2%2C%22include_lower%22%3Atrue%2C%22include_upper%22%3Afalse%7D%2C%22operator%22%3A%22range%22%7D%7D%5D%2C%22logical_operator%22%3A%22or%22%2C%22label%22%3A%22rcsb_entry_info.resolution_combined%22%7D%5D%2C%22logical_operator%22%3A%22and%22%7D%5D%2C%22logical_operator%22%3A%22and%22%2C%22label%22%3A%22text%22%7D%5D%2C%22logical_operator%22%3A%22and%22%7D%2C%22return_type%22%3A%22entry%22%2C%22request_options%22%3A%7B%22paginate%22%3A%7B%22start%22%3A0%2C%22rows%22%3A25%7D%2C%22results_content_type%22%3A%5B%22experimental%22%5D%2C%22sort%22%3A%5B%7B%22sort_by%22%3A%22score%22%2C%22direction%22%3A%22desc%22%7D%5D%2C%22scoring_strategy%22%3A%22combined%22%7D%2C%22request_info%22%3A%7B%22query_id%22%3A%22b1a8b4eaa1d164ac7c0d669775884147%22%7D%7D "Go RCSB"

### Total: 987 (2023)
   * Options (pdb 다운로드할 때)
      1. Organism: Homo sapiens
      2. Polymer Entity Type: Protein
      3. Resoultion (Å): ~ 2.0
    

<img src="pdb_2023_human_X_ray_Prot_resol2/img/pdb_download.png" width="60%" height="100%" title="PDB_Download_list"></img>

* 다운로드 방식
    - 옆에 batch 나온 것들을 이용
    - pdb 파일들을 복사하여 텍스트 파일을 만든 후 api로 다운 (선택)

#### 리간드 csv 다운로드 (2023)

<img src="pdb_2023_human_X_ray_Prot_resol2/img/ligand_report.png" width="60%" height="100%" title="ligand_csv_url"></img>

- csv 클릭하여 다운로드

#### pdb download (987개)

In [3]:
from get_pdb_from_rcsb_01 import get_pdb, gz_extract_pdb

In [4]:
pdbs_path = 'pdb_2023_human_X_ray_Prot_resol2/pdb_2023_human_X_ray_Prot_resol2.txt'
output = 'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb'
get_pdb(input=pdbs_path, output=output) # .gz 파일들로 다운로드됨.

Failed to download https://files.rcsb.org/download/7PA6.pdb.gz
Failed to download https://files.rcsb.org/download/7QNS.pdb.gz
Failed to download https://files.rcsb.org/download/7QNV.pdb.gz
Failed to download https://files.rcsb.org/download/7QO2.pdb.gz
Failed to download https://files.rcsb.org/download/7QOB.pdb.gz
Failed to download https://files.rcsb.org/download/7QRK.pdb.gz
Failed to download https://files.rcsb.org/download/7QSE.pdb.gz
Failed to download https://files.rcsb.org/download/7QSI.pdb.gz
Failed to download https://files.rcsb.org/download/7QUM.pdb.gz
Failed to download https://files.rcsb.org/download/7QYK.pdb.gz
Failed to download https://files.rcsb.org/download/7QZM.pdb.gz
Failed to download https://files.rcsb.org/download/7QZX.pdb.gz
Failed to download https://files.rcsb.org/download/7QZY.pdb.gz
Failed to download https://files.rcsb.org/download/7R1X.pdb.gz
Failed to download https://files.rcsb.org/download/7U4P.pdb.gz
Failed to download https://files.rcsb.org/download/7YWT

In [5]:
extract_path = 'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb'
gz_extract_pdb(input=extract_path) # .gz 파일들을 압축풀기



0


In [7]:
import pandas as pd
import numpy as np
import os

#### rcsb에서 가져온 pdb에서는 리간드의 분자량이 다양함 → 리간드 분자량 250~400인 pdb만 추출 

In [8]:
# rcsb 에서 다운로드한 ligand report 가져오기 (987개)
pdb_2023_human = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/01_pdb_lig_2023_human_X_ray_Prot_resol2.csv')
pdb_2023_human.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2387 entries, 0 to 2386
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Entry ID            987 non-null    object 
 1   Asym ID             2312 non-null   object 
 2   Auth Asym ID        2312 non-null   object 
 3   Ligand ID           2237 non-null   object 
 4   Ligand Formula      2312 non-null   object 
 5   Ligand MW           2308 non-null   float64
 6   Ligand Name         2312 non-null   object 
 7   InChI               2308 non-null   object 
 8   InChI Key           2308 non-null   object 
 9   Ligand SMILES       2312 non-null   object 
 10  Ligand of Interest  798 non-null    object 
dtypes: float64(1), object(10)
memory usage: 205.3+ KB


In [9]:
mw = pdb_2023_human[(pdb_2023_human['Ligand MW'] >= 250) & (pdb_2023_human['Ligand MW'] <= 400)]
# type(pdb[0])

pdb = []
for pdbid in mw['Entry ID'].tolist():
    # print(type(pdbid))
    if type(pdbid) == str:
        pdb.append(pdbid)
# len(mw['Entry ID']) == len(pdb) 

In [10]:
len(set(pdb))

241

In [11]:
mw_250_400 = mw[mw['Entry ID'].isin(pdb)]
mw_250_400.columns

Index(['Entry ID', 'Asym ID', 'Auth Asym ID', 'Ligand ID', 'Ligand Formula',
       'Ligand MW', 'Ligand Name', 'InChI', 'InChI Key', 'Ligand SMILES',
       'Ligand of Interest'],
      dtype='object')

In [12]:
info = ['Entry ID', 
        'Ligand ID', 'Ligand Name', 'Ligand Formula', 'Ligand SMILES', 'Ligand MW',   
        'InChI', 'InChI Key']
new_pdb = mw_250_400[info]

output = 'pdb_2023_human_X_ray_Prot_resol2/01_preprocessed_pdb.csv'
new_pdb.to_csv(output, index=False, index_label=False)

# 2. 리간드 분자량 250~400인 pdb만 추출한 csv 가져오기

In [13]:
import pandas as pd
import numpy as np
import pickle

from tqdm import tqdm
import logging

import pubchempy as pcp
from pubchempy import Compound, get_compounds, get

In [14]:
pdb_lig = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/01_preprocessed_pdb.csv')
pdb_lig.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 241 entries, 0 to 240
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Entry ID        241 non-null    object 
 1   Ligand ID       241 non-null    object 
 2   Ligand Name     241 non-null    object 
 3   Ligand Formula  241 non-null    object 
 4   Ligand SMILES   241 non-null    object 
 5   Ligand MW       241 non-null    float64
 6   InChI           241 non-null    object 
 7   InChI Key       241 non-null    object 
dtypes: float64(1), object(7)
memory usage: 15.2+ KB


In [15]:
print('Ligand ID:', len(pdb_lig['Ligand ID'].unique()), '\n',
      'InChI: ', len(pdb_lig['InChI'].unique()), '\n',
      'InChI Key: ', len(pdb_lig['InChI Key'].unique()), '\n',
      'Ligand Name: ', len(pdb_lig['Ligand Name'].unique()))

Ligand ID: 215 
 InChI:  215 
 InChI Key:  215 
 Ligand Name:  214


### ligand ID, ligand SMILES 다 시도해보았지만 완전히 못 찾기 때문에 cid,isomeric_smiles, iupac_name을 None값으로 저장함.

In [16]:
# pdb_lig['isomeric_smiles']=""
# pdb_lig['CID'] = ""
# pdb_lig['iupac_name'] = ""
# i = 0

isosmi = []
cids = []
iupac_names = []
r = 0
for inchi in tqdm(pdb_lig['InChI']):
      try:
            compound = get_compounds(inchi, namespace='inchi')
            if (compound[0].cid is None):
                  print("Since inch cannot find a compound, we will search by ligand name.", '\n',
                        inchi, compound, '\n',
                        'ligand name: ', pdb_lig['Ligand Name'][r])
                  compound = get_compounds(pdb_lig['Ligand Name'][r], namespace='name')
                  isomeric_smiles, cid, iupac_name = compound[0].isomeric_smiles, compound[0].cid, compound[0].iupac_name
                  isosmi.append(isomeric_smiles)
                  cids.append(cid)
                  iupac_names.append(iupac_name)
                  print('row: ', r+2, 'cid: ', compound[0].cid, '\n', '#'*150) # r+2는 그냥 체크하기 쉽게하기 위함.
                  r += 1
            else:
                  isomeric_smiles, cid, iupac_name = compound[0].isomeric_smiles, compound[0].cid, compound[0].iupac_name
                  r += 1
                  
                  isosmi.append(isomeric_smiles)
                  cids.append(cid)
                  iupac_names.append(iupac_name)
            
            # Ignore SettingWithCopy Warning 
            # pd.set_option('mode.chained_assignment', None)
            
            # make new column & insert values
            # pdb_lig['isomeric_smiles'][i] = isomeric_smiles
            # pdb_lig['CID'][i] = cid
            # pdb_lig['iupac_name'][i] = iupac_name
            # i += 1
            
      except IndexError as ie:
            print(r+2, 'IndexError: ', ie) # r+2는 그냥 체크하기 쉽게하기 위함.
            r += 1
            isosmi.append(None)
            cids.append(None)
            iupac_names.append(None)
            
            # isosmi.append('0')
            # cids.append(0)
            # iupac_names.append('0')
      except TypeError as t:
            print(r+2, 'TypeError: ',t) # r+2는 그냥 체크하기 쉽게하기 위함.
            r += 1
            isosmi.append(None)
            cids.append(None)
            iupac_names.append(None)
            
            # isosmi.append('0')
            # cids.append(0)
            # iupac_names.append('0')
            
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_isosmi.pickle', 'wb') as pi:
      pickle.dump(isosmi, pi)
      
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_cids.pickle', 'wb') as pc:
      pickle.dump(cids, pc)
      
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_iupac_names.pickle', 'wb') as pn:
      pickle.dump(iupac_names, pn)

  5%|▍         | 11/241 [00:25<07:50,  2.05s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C15H21N3O3S/c19-14(13-4-2-10-22-13)17-5-7-18(8-6-17)15(20)16-11-12-3-1-9-21-12/h2,4,10,12H,1,3,5-9,11H2,(H,16,20)/t12-/m0/s1 [Compound()] 
 ligand name:  N-{[(2S)-oxolan-2-yl]methyl}-4-(thiophene-2-carbonyl)piperazine-1-carboxamide


  5%|▍         | 12/241 [00:28<08:30,  2.23s/it]

row:  13 cid:  167530290 
 ######################################################################################################################################################


  7%|▋         | 18/241 [00:38<07:02,  1.90s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C15H23N3O3/c1-3-4-7-16-15(20)18-9-8-17(11-12(18)2)14(19)13-6-5-10-21-13/h5-6,10,12H,3-4,7-9,11H2,1-2H3,(H,16,20)/t12-/m1/s1 [Compound()] 
 ligand name:  (2R)-N-butyl-4-(furan-2-carbonyl)-2-methylpiperazine-1-carboxamide


  8%|▊         | 19/241 [00:42<08:23,  2.27s/it]

row:  20 cid:  167530294 
 ######################################################################################################################################################


  9%|▉         | 22/241 [00:47<07:17,  2.00s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C14H21N3O3/c1-10(2)15-14(19)16-6-7-17(11(3)9-16)13(18)12-5-4-8-20-12/h4-5,8,10-11H,6-7,9H2,1-3H3,(H,15,19)/t11-/m1/s1 [Compound()] 
 ligand name:  (3R)-4-(furan-2-carbonyl)-3-methyl-N-(propan-2-yl)piperazine-1-carboxamide


 10%|▉         | 23/241 [00:49<07:51,  2.16s/it]

row:  24 cid:  167530295 
 ######################################################################################################################################################


 10%|▉         | 24/241 [00:51<07:07,  1.97s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C15H21N3O3/c1-11-10-17(14(19)13-3-2-8-21-13)6-7-18(11)15(20)16-9-12-4-5-12/h2-3,8,11-12H,4-7,9-10H2,1H3,(H,16,20)/t11-/m0/s1 [Compound()] 
 ligand name:  (2S)-N-(cyclopropylmethyl)-4-(furan-2-carbonyl)-2-methylpiperazine-1-carboxamide


 10%|█         | 25/241 [00:54<08:00,  2.22s/it]

row:  26 cid:  167530296 
 ######################################################################################################################################################


 11%|█         | 27/241 [00:57<07:11,  2.01s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C19H27N3O2S2/c1-13(19(2,3)4)12-20-18(24)22-8-6-21(7-9-22)17(23)16-11-15-14(26-16)5-10-25-15/h5,10-11,13H,6-9,12H2,1-4H3,(H,20,24)/t13-/m1/s1 [Compound()] 
 ligand name:  4-(thieno[3,2-b]thiophene-2-carbonyl)-N-[(2S)-2,3,3-trimethylbutyl]piperazine-1-carboxamide


 12%|█▏        | 28/241 [01:00<07:31,  2.12s/it]

29 IndexError:  list index out of range


 12%|█▏        | 30/241 [01:03<07:10,  2.04s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H21N3O4/c1-13-11-20(17(22)14-7-10-25-12-14)8-9-21(13)18(23)19-15-3-5-16(24-2)6-4-15/h3-7,10,12-13H,8-9,11H2,1-2H3,(H,19,23)/t13-/m1/s1 [Compound()] 
 ligand name:  (2R)-4-(furan-3-carbonyl)-N-(4-methoxyphenyl)-2-methylpiperazine-1-carboxamide


 13%|█▎        | 31/241 [01:06<08:04,  2.31s/it]

row:  32 cid:  167530298 
 ######################################################################################################################################################


 13%|█▎        | 32/241 [01:08<07:16,  2.09s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C13H16F3N3O3/c1-9-7-18(11(20)10-3-2-6-22-10)4-5-19(9)12(21)17-8-13(14,15)16/h2-3,6,9H,4-5,7-8H2,1H3,(H,17,21)/t9-/m0/s1 [Compound()] 
 ligand name:  (2S)-4-(furan-2-carbonyl)-2-methyl-N-(2,2,2-trifluoroethyl)piperazine-1-carboxamide


 14%|█▎        | 33/241 [01:11<07:56,  2.29s/it]

row:  34 cid:  167530299 
 ######################################################################################################################################################


 19%|█▉        | 46/241 [01:33<05:06,  1.57s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C11H11Cl3O2S/c1-5(2)10(11(15)16)17-9-4-7(13)6(12)3-8(9)14/h3-5,10H,1-2H3,(H,15,16)/t10-/m1/s1 [Compound()] 
 ligand name:  (2R)-3-methyl-2-[(2,4,5-trichlorophenyl)sulfanyl]butanoic acid


 20%|█▉        | 47/241 [01:35<05:59,  1.85s/it]

48 IndexError:  list index out of range


 20%|█▉        | 48/241 [01:38<06:35,  2.05s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C17H15ClO2/c18-14-8-6-11(7-9-14)13-5-4-12-2-1-3-15(17(19)20)16(12)10-13/h4-10,15H,1-3H2,(H,19,20)/t15-/m1/s1 [Compound()] 
 ligand name:  (1R)-7-(4-chlorophenyl)-1,2,3,4-tetrahydronaphthalene-1-carboxylic acid


 20%|██        | 49/241 [01:40<06:40,  2.09s/it]

50 IndexError:  list index out of range


 23%|██▎       | 55/241 [01:51<05:37,  1.82s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C13H11ClO3/c14-10-3-1-8(2-4-10)9-5-12(16)11(7-15)13(17)6-9/h1-4,7,9,15H,5-6H2/b11-7-/t9-/m0/s1 [Compound()] 
 ligand name:  (2seqTrans,5S)-5-(4-chlorophenyl)-2-(hydroxymethylidene)cyclohexane-1,3-dione


 23%|██▎       | 56/241 [01:53<06:12,  2.01s/it]

row:  57 cid:  168300843 
 ######################################################################################################################################################


 24%|██▍       | 59/241 [01:58<05:13,  1.72s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H16ClNO2/c19-15-7-6-14(12-20-9-8-16(21)11-18(20)22)17(10-15)13-4-2-1-3-5-13/h1-7,10-11,21H,8-9,12H2 [Compound()] 
 ligand name:  1-[(5-chloro[1,1'-biphenyl]-2-yl)methyl]-4-hydroxy-5,6-dihydropyridin-2(1H)-one


 25%|██▍       | 60/241 [02:01<06:25,  2.13s/it]

row:  61 cid:  168300848 
 ######################################################################################################################################################


 26%|██▌       | 63/241 [02:06<05:21,  1.81s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C19H14O3/c20-12-9-13(22-19(21)10-12)11-18-16-7-3-1-5-14(16)15-6-2-4-8-17(15)18/h1-8,10-11,13,20H,9H2/t13-/m0/s1 [Compound()] 
 ligand name:  (6S)-6-{[(4aM)-9H-fluoren-9-ylidene]methyl}-4-hydroxy-5,6-dihydro-2H-pyran-2-one


 27%|██▋       | 64/241 [02:08<05:44,  1.95s/it]

row:  65 cid:  168300852 
 ######################################################################################################################################################


 27%|██▋       | 65/241 [02:10<05:23,  1.84s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H17BrN2OS/c19-15-8-6-7-14(13-15)17(22)21-11-4-5-12-23-18(21)20-16-9-2-1-3-10-16/h1-3,6-10,13H,4-5,11-12H2/b20-18- [Compound()] 
 ligand name:  (3-bromophenyl)[(2Z)-2-(phenylimino)-1,3-thiazepan-3-yl]methanone


 27%|██▋       | 66/241 [02:12<06:06,  2.10s/it]

row:  67 cid:  168300855 
 ######################################################################################################################################################


 33%|███▎      | 80/241 [02:36<04:17,  1.60s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C15H21N4OS2/c1-2-3-4-5-9-22-13-17-14(20)19(15(21)18-13)11-12-7-6-8-16-10-12/h6-8,10H,2-5,9,11H2,1H3,(H2,17,18,20,21) [Compound()] 
 ligand name:  4-hexylsulfanyl-1-(pyridin-3-ylmethyl)-6-sulfanyl-1$l^{4},3,5-triazacyclohexa-1,3,5-trien-2-ol


 34%|███▎      | 81/241 [02:38<05:11,  1.95s/it]

row:  82 cid:  168300873 
 ######################################################################################################################################################
Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C24H29N5/c1-16-19(15-17-11-7-5-8-12-17)25-22(24(2,3)4)21(23-26-28-29-27-23)20(16)18-13-9-6-10-14-18/h5-14,20-22,25H,15H2,1-4H3,(H,26,27,28,29)/t20-,21+,22-/m1/s1 [Compound()] 
 ligand name:  (2R,3S,4S)-6-benzyl-2-tert-butyl-5-methyl-4-phenyl-3-(1H-tetrazol-5-yl)-1,2,3,4-tetrahydropyridine


 34%|███▍      | 82/241 [02:41<05:44,  2.17s/it]

row:  83 cid:  168300874 
 ######################################################################################################################################################
Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H16N4/c19-18-20-11-14-10-13-8-4-5-9-15(13)16(21-17(14)22-18)12-6-2-1-3-7-12/h1-9,11,16H,10H2,(H3,19,20,21,22)/t16-/m0/s1 [Compound()] 
 ligand name:  (10S)-10-phenyl-10,11-dihydro-5H-pyrimido[4,5-c][2]benzazepin-2-amine


 34%|███▍      | 83/241 [02:44<06:05,  2.31s/it]

row:  84 cid:  168300876 
 ######################################################################################################################################################


 35%|███▌      | 85/241 [02:47<05:14,  2.02s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C19H21ClO2S2/c1-14-5-9-17(10-6-14)24-18(19(21)22)4-2-3-13-23-16-11-7-15(20)8-12-16/h5-12,18H,2-4,13H2,1H3,(H,21,22)/t18-/m0/s1 [Compound()] 
 ligand name:  (2S)-6-[(4-chlorophenyl)sulfanyl]-2-[(4-methylphenyl)sulfanyl]hexanoic acid


 36%|███▌      | 86/241 [02:50<05:38,  2.18s/it]

row:  87 cid:  168300881 
 ######################################################################################################################################################


 38%|███▊      | 92/241 [02:59<03:58,  1.60s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C16H16ClNO3/c1-7-10-6-11(12(7)16(19)20)13-14(18-21-15(10)13)8-2-4-9(17)5-3-8/h2-5,7,10-13,15H,6H2,1H3,(H,19,20)/t7-,10+,11-,12+,13+,15+/m0/s1 [Compound()] 
 ligand name:  (3aR,4R,5R,6S,7R,7aR)-3-(4-chlorophenyl)-6-methyl-3a,4,5,6,7,7a-hexahydro-4,7-methano-1,2-benzoxazole-5-carboxylic acid


 39%|███▊      | 93/241 [03:02<04:42,  1.91s/it]

row:  94 cid:  168300885 
 ######################################################################################################################################################


 39%|███▉      | 94/241 [03:03<04:25,  1.81s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H16N2O3S/c21-15(22)10-11-24-18(19-12-6-2-1-3-7-12)16-13-8-4-5-9-14(13)20-17(16)23/h1-9,19H,10-11H2,(H,20,23)(H,21,22)/b18-16+ [Compound()] 
 ligand name:  3-{[(E)-anilino(2-oxo-1,2-dihydro-3H-indol-3-ylidene)methyl]sulfanyl}propanoic acid


 39%|███▉      | 95/241 [03:06<05:15,  2.16s/it]

row:  96 cid:  168300889 
 ######################################################################################################################################################


 40%|███▉      | 96/241 [03:08<04:41,  1.94s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C12H10ClO3S/c13-9-1-3-10(4-2-9)16-11-5-7-12(8-6-11)17(14)15/h1-8,14-15H/q+1 [Compound()] 
 ligand name:  1-[bis(oxidanyl)-$l^{3}-sulfanyl]-4-(4-chloranylphenoxy)benzene


 40%|████      | 97/241 [03:10<05:08,  2.14s/it]

row:  98 cid:  70302558 
 ######################################################################################################################################################


 42%|████▏     | 102/241 [03:19<04:16,  1.85s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C12H18N2O2S/c1-2-17-12-13-10(15)8-11(16)14(12)9-6-4-3-5-7-9/h9H,2-8H2,1H3 [Compound()] 
 ligand name:  1-cyclohexyl-2-(ethylsulfanyl)pyrimidine-4,6(1H,5H)-dione


 43%|████▎     | 103/241 [03:22<05:01,  2.18s/it]

row:  104 cid:  1225058 
 ######################################################################################################################################################


 54%|█████▎    | 129/241 [04:08<03:13,  1.73s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C16H20BrNO3/c1-2-3-4-5-6-11-14(16(20)21)12-9-10(17)7-8-13(12)18-15(11)19/h7-9,11,14H,2-6H2,1H3,(H,18,19)(H,20,21)/t11-,14+/m0/s1 [Compound()] 
 ligand name:  (3S,4R)-6-bromo-3-hexyl-2-oxo-1,2,3,4-tetrahydroquinoline-4-carboxylic acid


 54%|█████▍    | 130/241 [04:10<03:36,  1.95s/it]

row:  131 cid:  168300916 
 ######################################################################################################################################################


 59%|█████▊    | 141/241 [04:31<03:04,  1.84s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C19H23N3O4S/c1-10-20-17(26-22-10)15-13-8-4-5-9-14(13)27-18(15)21-16(23)11-6-2-3-7-12(11)19(24)25/h11-12H,2-9H2,1H3,(H,21,23)(H,24,25)/t11-,12+/m1/s1 [Compound()] 
 ligand name:  (1R,2R)-2-{[(3M)-3-(3-methyl-1,2,4-oxadiazol-5-yl)-4,5,6,7-tetrahydro-1-benzothiophen-2-yl]carbamoyl}cyclohexane-1-carboxylic acid


 59%|█████▉    | 142/241 [04:34<03:45,  2.28s/it]

row:  143 cid:  168300926 
 ######################################################################################################################################################


 61%|██████    | 147/241 [04:43<02:59,  1.91s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C18H22N2O4/c1-5-10-18(12(2)3)15(21)19-17(23)20(16(18)22)11-13-6-8-14(24-4)9-7-13/h5-9,12H,1,10-11H2,2-4H3,(H,19,21,23)/t18-/m0/s1 [Compound()] 
 ligand name:  (5S)-1-[(4-methoxyphenyl)methyl]-5-(propan-2-yl)-5-(prop-2-en-1-yl)-1,3-diazinane-2,4,6-trione


 61%|██████▏   | 148/241 [04:45<03:16,  2.11s/it]

149 IndexError:  list index out of range
Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C14H16N2O3/c1-9-5-6-13(19-9)11-8-12(14-4-3-7-18-14)16(15-11)10(2)17/h3-7,11-12,15H,8H2,1-2H3/t11-,12-/m1/s1 [Compound()] 
 ligand name:  1-[(3R,5R)-5-(furan-2-yl)-3-(5-methylfuran-2-yl)pyrazolidin-1-yl]ethan-1-one


 62%|██████▏   | 149/241 [04:48<03:36,  2.35s/it]

150 IndexError:  list index out of range


 85%|████████▍ | 204/241 [06:27<01:01,  1.66s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C16H11IN2O/c17-15-10-14(7-6-13(15)11-20)19-9-8-18-16(19)12-4-2-1-3-5-12/h1-11H [Compound()] 
 ligand name:  2-iodanyl-4-(2-phenylimidazol-1-yl)benzaldehyde


 85%|████████▌ | 205/241 [06:30<01:15,  2.08s/it]

row:  206 cid:  168300966 
 ######################################################################################################################################################


 98%|█████████▊| 236/241 [07:24<00:08,  1.64s/it]

Since inch cannot find a compound, we will search by ligand name. 
 InChI=1S/C20H21N5O2/c1-12-10-18(16(22)11-19(12)26)25-24-14-8-6-13(7-9-14)20(27)23-17-5-3-2-4-15(17)21/h2-11,24-26H,21-22H2,1H3,(H,23,27) [Compound()] 
 ligand name:  ~{N}-(2-aminophenyl)-4-[2-(2-azanyl-5-methyl-4-oxidanyl-phenyl)hydrazinyl]benzamide


 98%|█████████▊| 237/241 [07:27<00:07,  1.89s/it]

row:  238 cid:  168355726 
 ######################################################################################################################################################


100%|██████████| 241/241 [07:33<00:00,  1.88s/it]


In [17]:
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_isosmi.pickle', 'rb') as pi:
      isosmi = pickle.load(pi)
      
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_cids.pickle', 'rb') as pc:
      cids = pickle.load(pc)
      
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_iupac_names.pickle', 'rb') as pn:
      iupac_names = pickle.load(pn)

In [19]:
pdb_lig['isomeric_smiles']=""
pdb_lig['CID'] = ""
pdb_lig['iupac_name'] = ""
i = 0

for isomeric_smiles, cid, iupac_name in zip(isosmi, cids, iupac_names):
    # make new column & insert values
    pdb_lig['isomeric_smiles'][i] = isomeric_smiles
    pdb_lig['CID'][i] = cid
    pdb_lig['iupac_name'][i] = iupac_name
    i += 1

/tmp/ipykernel_9760/3081893019.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pdb_lig['isomeric_smiles'][i] = isomeric_smiles
/tmp/ipykernel_9760/3081893019.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pdb_lig['CID'][i] = cid
/tmp/ipykernel_9760/3081893019.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pdb_lig['iupac_name'][i] = iupac_name
/tmp/ipykernel_9760/3081893019.py:8: SettingWithCopyWarning: 
A value is trying

In [20]:
pdb_lig.to_csv('pdb_2023_human_X_ray_Prot_resol2/02_final_preprocessed_pdb.csv', index=False, index_label=False)

# 3. pubchempy 이용하여 3d 유무 확인하여 분리하기 


In [45]:
import pandas as pd

from rdkit import Chem
from rdkit.Chem import PandasTools, rdMolDescriptors

* LID, L_Name, L_MW, L_Formula, SMILES, InChI, InChI Key는 rcsb에서 가져온 것
* isomeric_smiles, CID, iupac_name: pubchempy api 이용하여 가져온 것

In [21]:
# path = 'pdb_2023_human_X_ray_Prot_resol2/03_final_preprocessed_pdb_none.csv'
path = 'pdb_2023_human_X_ray_Prot_resol2/02_final_preprocessed_pdb.csv'
final_pdb_lig = pd.read_csv(path)
final_pdb_lig = final_pdb_lig.set_axis(['PDB', 'LID', 'L_Name', 'L_Formula',
                        'SMILES', 'L_MW', 'InChI', 'InChI Key', 'isomeric_smiles',
                        'CID', 'iupac_name'], axis=1)
len(final_pdb_lig)

241

In [22]:
col = ['PDB', 'LID',  'CID', 'L_MW', 'iupac_name', 'isomeric_smiles', 'InChI', 'InChI Key']
final_pdb_lig = final_pdb_lig[col].dropna(axis=0)
final_pdb_lig.reset_index(drop=True)
len(final_pdb_lig)

236

In [23]:
final_pdb_lig.to_csv('pdb_2023_human_X_ray_Prot_resol2/03_final_pdb_lig.csv', index=False, index_label=False)

* Protein Structure Describe
    * protein 정보에서 위에서 처리하였던 리간드의 pdb를 따로 저장하기 위함

In [27]:
final_pdb_lig = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/03_final_pdb_lig.csv')
final_pdb_lig.head()

PDB  LID      CID     L_MW   
0  7FRW  FBP  10267.0  340.116  \
1  7FRX  FBP  10267.0  340.116   
2  7FRY  FBP  10267.0  340.116   
3  7FS1  FBP  10267.0  340.116   
4  7FS3  FBP  10267.0  340.116   

                                          iupac_name   
0  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...  \
1  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
2  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
3  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
4  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   

                                     isomeric_smiles   
0  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...  \
1  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
2  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
3  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
4  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   

                                               InChI   
0  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...  \
1  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
2  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
3  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
4  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   

                     InChI Key  
0  RNBGYGVWRKECFJ-ARQDHWQXSA-N  
1  RNBGYGVWRKECFJ-ARQDHWQXSA-N  
2  RNBGYGVWRKECFJ-ARQDHWQXSA-N  
3  RNBGYGVWRKECFJ-ARQDHWQXSA-N  
4  RNBGYGVWRKECFJ-ARQDHWQXSA-N

- protein structure information

<img src="pdb_2023_human_X_ray_Prot_resol2/img/pdb_strucuture_download.png" width="60%" height="100%" title="ligand_csv_url"></img>

In [28]:
prot_info = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/03_pdb_2023_info.csv')

prot_info.columns = prot_info.iloc[0]

prot_info  = prot_info[1:]

len(prot_info)

1013

In [29]:
pdb_list = final_pdb_lig['PDB'].tolist()

extract_pdb = prot_info[prot_info['Entry ID'].isin(pdb_list)]
extract_pdb.to_csv('pdb_2023_human_X_ray_Prot_resol2/03_prot_info_extract_pdb.csv', index=False, index_label=False)

In [30]:
prot_info = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/03_prot_info_extract_pdb.csv')
prot_info.columns = ['PDB', 'Structure Title', 'Structure Determination Methodology',
       'Resolution (Å)', 'Molecular Weight per Deposited Model',
       'Total Number of Polymer Residues per Deposited Model',
       'Number of Non-Hydrogen Atoms per Deposited Model',
       'Experimental Method', 'Release Date', 'Stucture Keywords',
       'Structure Author', 'Unnamed: 11']

In [31]:
len(final_pdb_lig) == len(prot_info)

True

* pdb 행이 같은지 확인

In [32]:
for pl, pi in zip(final_pdb_lig['PDB'],prot_info['PDB']):
    if pl != pi:
        print('False')
    # else:
    #     print('True')

* pd.merge() 이용하여 합치기

In [33]:
final_pdb_lig_prot_info = pd.merge(pdb_lig, prot_info, how='left', on='PDB')
final_pdb_lig_prot_info.to_csv('pdb_2023_human_X_ray_Prot_resol2/03_final_pdb_lig_prot_info.csv', index=False, index_label=False)

In [34]:
final_pdb_lig_prot_info = pd.read_csv('pdb_2023_human_X_ray_Prot_resol2/03_final_pdb_lig_prot_info.csv')
final_pdb_lig_prot_info = final_pdb_lig_prot_info.drop('Unnamed: 11', axis=1) # none 값인 컬럼 제거
final_pdb_lig_prot_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 236 entries, 0 to 235
Data columns (total 18 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   PDB                                                   236 non-null    object 
 1   LID                                                   236 non-null    object 
 2   CID                                                   236 non-null    float64
 3   L_MW                                                  236 non-null    float64
 4   iupac_name                                            236 non-null    object 
 5   isomeric_smiles                                       236 non-null    object 
 6   InChI                                                 236 non-null    object 
 7   InChI Key                                             236 non-null    object 
 8   Structure Title                                       236 no

In [35]:
final_pdb_lig_prot_info = final_pdb_lig_prot_info.astype({'CID':'int'})
final_pdb_lig_prot_info.head()

PDB  LID    CID     L_MW   
0  7FRW  FBP  10267  340.116  \
1  7FRX  FBP  10267  340.116   
2  7FRY  FBP  10267  340.116   
3  7FS1  FBP  10267  340.116   
4  7FS3  FBP  10267  340.116   

                                          iupac_name   
0  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...  \
1  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
2  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
3  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   
4  [(2R,3S,4S,5R)-2,3,4-trihydroxy-5-(phosphonoox...   

                                     isomeric_smiles   
0  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...  \
1  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
2  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
3  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   
4  C([C@@H]1[C@H]([C@@H]([C@](O1)(COP(=O)(O)O)O)O...   

                                               InChI   
0  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...  \
1  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
2  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
3  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   
4  InChI=1S/C6H14O12P2/c7-4-3(1-16-19(10,11)12)18...   

                     InChI Key   
0  RNBGYGVWRKECFJ-ARQDHWQXSA-N  \
1  RNBGYGVWRKECFJ-ARQDHWQXSA-N   
2  RNBGYGVWRKECFJ-ARQDHWQXSA-N   
3  RNBGYGVWRKECFJ-ARQDHWQXSA-N   
4  RNBGYGVWRKECFJ-ARQDHWQXSA-N   

                                     Structure Title   
0  Structure of liver pyruvate kinase in complex ...  \
1  Structure of liver pyruvate kinase in complex ...   
2  Structure of liver pyruvate kinase in complex ...   
3  Structure of liver pyruvate kinase in complex ...   
4  Structure of liver pyruvate kinase in complex ...   

  Structure Determination Methodology  Resolution (Å)   
0                        experimental           1.737  \
1                        experimental           1.847   
2                        experimental           1.958   
3                        experimental           1.863   
4                        experimental           1.655   

   Molecular Weight per Deposited Model   
0                                392.46  \
1                                392.02   
2                                392.08   
3                                392.45   
4                                394.56   

   Total Number of Polymer Residues per Deposited Model   
0                                               3576     \
1                                               3576      
2                                               3576      
3                                               3576      
4                                               3576      

   Number of Non-Hydrogen Atoms per Deposited Model Experimental Method   
0                                             29188   X-RAY DIFFRACTION  \
1                                             29119   X-RAY DIFFRACTION   
2                                             28032   X-RAY DIFFRACTION   
3                                             28898   X-RAY DIFFRACTION   
4                                             29766   X-RAY DIFFRACTION   

  Release Date Stucture Keywords   
0   2023-04-12       TRANSFERASE  \
1   2023-04-12       TRANSFERASE   
2   2023-04-12       TRANSFERASE   
3   2023-04-12       TRANSFERASE   
4   2023-04-12       TRANSFERASE   

                                    Structure Author  
0  Lulla, A., Nilsson, O., Brear, P., Nain-Perez,...  
1  Lulla, A., Nilsson, O., Brear, P., Nain-Perez,...  
2  Lulla, A., Nilsson, O., Brear, P., Nain-Perez,...  
3  Lulla, A., Nilsson, O., Brear, P., Nain-Perez,...  
4  Lulla, A., Nilsson, O., Brear, P., Nain-Perez,...

In [36]:
exist_3d = []
for isosmi in final_pdb_lig_prot_info['isomeric_smiles']:
    c = get_compounds(isosmi, namespace='smiles', record_type='3d')
    if len(c) < 1:
        print('empty')
    else:
        print(c[0].cid)
        exist_3d.append(c[0].cid)
        
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_exist_3d.pickle', 'wb') as c3d:
      pickle.dump(exist_3d, c3d)

10267
10267
10267
10267
10267
10267
10267
10267
10267
2989118
3583620
167530290
47127361
47128378
166217349
166309888
4415556
47128052
167530294
2554172
110491926
166219265
167530295
166181138
167530296
empty
166180534
3719721
166181210
167530298
27516341
167530299
5182372
167530300
108812029
47128041
166222812
166181152
47110368
166184745
1752219
82304928
13409295
43152220
empty
756974
985
empty
774837
empty
23613928
2806001
empty
2236831
90673045
empty
empty
empty
empty
135764101
empty
empty
empty
826598
empty
73296406
empty
empty
83340684
90042439
90043049
empty
2821498
73441851
empty
empty
empty
empty
empty
empty
61190858
73296263
empty
empty
2732367
11149855
87441312
empty
empty
empty
empty
empty
92179184
empty
empty
empty
18283071
empty
732030
1225058
697679
empty
906397
empty
empty
empty
61190858
empty
71556425
2807687
15097579
15680426
58447996
610479
1216100
empty
empty
24885077
empty
4430408
731146
10520703
empty
empty
empty
empty
empty
empty
empty
118589179
empty
empty
empty

In [37]:
with open('pdb_2023_human_X_ray_Prot_resol2/pubchem_exist_3d.pickle', 'rb') as c3d:
      exist_3d = pickle.load(c3d)

In [40]:
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist'
os.makedirs(output, exist_ok=True)

final_pdb_lig_prot_info_3d_exist = final_pdb_lig_prot_info[final_pdb_lig_prot_info['CID'].isin(exist_3d)]
final_pdb_lig_prot_info_3d_exist.to_csv(f"{output}/final_rcsb_pubchempy_2023_3d_exist.csv", index=False, index_label=False)

In [42]:
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty'
os.makedirs(output, exist_ok=True)

final_pdb_lig_prot_info_3d_empty = final_pdb_lig_prot_info[~final_pdb_lig_prot_info['CID'].isin(exist_3d)]
final_pdb_lig_prot_info_3d_empty.to_csv(f"{output}/final_pdb_lig_prot_info_3d_empty.csv", index=False, index_label=False)

# 4. Pandastools 이용하여 그림 추가하기

In [7]:
import pandas as pd

from rdkit import Chem
from rdkit.Chem import PandasTools

In [44]:
path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/final_rcsb_pubchempy_2023_3d_exist.csv'
final_pdb_lig_prot_info_3d_exist = pd.read_csv(path)
len(final_pdb_lig_prot_info_3d_exist)

156

In [45]:
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/'
PandasTools.AddMoleculeColumnToFrame(final_pdb_lig_prot_info_3d_exist, smilesCol='isomeric_smiles')
PandasTools.SaveXlsxFromFrame(final_pdb_lig_prot_info_3d_exist, output + 'final_rcsb_pubchempy_2023_3d_exist.xls' , size=( 150 , 150 ))
PandasTools.WriteSDF(final_pdb_lig_prot_info_3d_exist,  output + 'final_rcsb_pubchempy_2023_3d_exist.sdf', properties=list(final_pdb_lig_prot_info_3d_exist.columns))

In [46]:
path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/final_pdb_lig_prot_info_3d_empty.csv'
final_pdb_lig_prot_info_3d_empty = pd.read_csv(path)
len(final_pdb_lig_prot_info_3d_empty)

80

In [47]:
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/'
PandasTools.AddMoleculeColumnToFrame(final_pdb_lig_prot_info_3d_empty, smilesCol='isomeric_smiles')
PandasTools.SaveXlsxFromFrame(final_pdb_lig_prot_info_3d_empty, output + 'final_pdb_lig_prot_info_3d_empty.xls' , size=( 150 , 150 ))
PandasTools.WriteSDF(final_pdb_lig_prot_info_3d_empty,  output + 'final_pdb_lig_prot_info_3d_empty.sdf', properties=list(final_pdb_lig_prot_info_3d_empty.columns))

# 5. Resolution 최소인 것 추출하기

In [1]:
import pandas as pd

from rdkit import Chem
from rdkit.Chem import PandasTools

In [4]:
path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/final_rcsb_pubchempy_2023_3d_exist.csv'
final_pdb_lig_prot_info_3d_exist = pd.read_csv(path)


path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/final_pdb_lig_prot_info_3d_empty.csv'
final_pdb_lig_prot_info_3d_empty = pd.read_csv(path)

print('Length of 3d_exist:', len(final_pdb_lig_prot_info_3d_exist), '\n',
      'Length of 3d_empty:',len(final_pdb_lig_prot_info_3d_empty))

Length of 3d_exist: 156 
 Length of 3d_empty: 80


* 저자 이름 개수

In [3]:
exist_authors = final_pdb_lig_prot_info_3d_exist['Structure Author'].unique().tolist()

exist_resol = dict()
for author in exist_authors:
    author_df = final_pdb_lig_prot_info_3d_exist[final_pdb_lig_prot_info_3d_exist['Structure Author']==author]
    if author not in exist_resol.keys():
        exist_resol[author]= min(author_df['Resolution (Å)'])
    else:
        continue
    
print('Number of exist_Author: ', len(final_pdb_lig_prot_info_3d_exist['Structure Author'].unique()))

empty_authors = final_pdb_lig_prot_info_3d_empty['Structure Author'].unique().tolist()

empty_resol = dict()
for author in empty_authors:
    author_df = final_pdb_lig_prot_info_3d_empty[final_pdb_lig_prot_info_3d_empty['Structure Author']==author]
    if author not in empty_resol.keys():
        empty_resol[author]= min(author_df['Resolution (Å)'])
    else:
        continue
    
print('Number of empty_Author: ', len(final_pdb_lig_prot_info_3d_empty['Structure Author'].unique()))

Number of exist_Author:  69
Number of empty_Author:  48


In [4]:
exist_min_pdb = final_pdb_lig_prot_info_3d_exist[final_pdb_lig_prot_info_3d_exist['Resolution (Å)'].isin(exist_resol.values())]
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/resol_min_pdb_3d_exist.csv'
exist_min_pdb.to_csv(output, index=False, index_label=False)

empty_min_pdb = final_pdb_lig_prot_info_3d_empty[final_pdb_lig_prot_info_3d_empty['Resolution (Å)'].isin(empty_resol.values())]
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/resol_min_pdb_3d_empty.csv'
empty_min_pdb.to_csv(output, index=False, index_label=False)

In [2]:
exist_path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/resol_min_pdb_3d_exist.csv'
exist_resol_min = pd.read_csv(exist_path)


empty_path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/resol_min_pdb_3d_empty.csv'
empty_resol_min = pd.read_csv(empty_path)


print('Length of exist_resol_min:', len(exist_resol_min), '\n',
      'Length of empty_resol_min:',len(empty_resol_min))

Length of exist_resol_min: 116 
 Length of empty_resol_min: 56


In [3]:
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/'

PandasTools.AddMoleculeColumnToFrame(exist_resol_min, smilesCol='isomeric_smiles')
PandasTools.SaveXlsxFromFrame(exist_resol_min, output + 'resol_min_pdb_3d_exist.xls' , size=( 150 , 150 ))
PandasTools.WriteSDF(exist_resol_min,  output + 'resol_min_pdb_3d_exist.sdf', properties=list(exist_resol_min.columns))

output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_empty/'
PandasTools.AddMoleculeColumnToFrame(empty_resol_min, smilesCol='isomeric_smiles')
PandasTools.SaveXlsxFromFrame(empty_resol_min, output + 'resol_min_pdb_3d_empty.xls' , size=( 150 , 150 ))
PandasTools.WriteSDF(empty_resol_min,  output + 'resol_min_pdb_3d_empty.sdf', properties=list(empty_resol_min.columns))

# 6. pubchem api 이용하여 3d 다운하기

In [4]:
import pandas as pd 
import os
import subprocess
from tqdm import tqdm

import requests
import urllib
import urllib.request
from urllib.error import HTTPError, URLError

from rdkit import Chem
from rdkit.Chem import PandasTools

In [5]:
path = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/resol_min_pdb_3d_exist.sdf'
sdf = PandasTools.LoadSDF(path)

## api 이용 예

#### 1. Find 3d ligands for cid 
- cid: 2244
    - https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/2244/conformers/TXT?record_type=3d

#### 2. Download 3d conformer id + cid
- id: 09FC4F3A00000016
- cid: 167530298
    - https://pubchem.ncbi.nlm.nih.gov/rest/pug/conformers/09FC4F3A00000016/SDF?response_type=save&response_basename=Conformer3D_COMPOUND_CID_167530298_09FC4F3A00000016

#### 3. Download 3d conformer isomeric_smiles + cid
- cid: 167530298
- iso_smi: 'C1CN(CCN1C(=O)C2=CC=CO2)C(=O)NCC3=CC4=C(C=C3)OCO4'
    - https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{C1CN(CCN1C(=O)C2=CC=CO2)C(=O)NCC3=CC4=C(C=C3)OCO4}/SDF?record_type=3d&response_type=save&response_basename=Conformer3D_COMPOUND_CID_{167530298}

# 3d 여러개 다운로드할 때

In [8]:
# pdb, cid, lid of sdf
exist_3d_pdb = sdf['PDB'].tolist()
exist_3d_cid = sdf['CID'].tolist()
exist_3d_lid = sdf['LID'].tolist()
exist_3d_isosmi = sdf['isomeric_smiles'].tolist()

# original pdb path
ori_pdb_path = 'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb'
ori_pdb = os.listdir(ori_pdb_path)

# save pdb & 3d conformer
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_multiple'

for digit, cid, lid, iso_smi in zip(exist_3d_pdb, exist_3d_cid, exist_3d_lid, exist_3d_isosmi):
    os.makedirs(output + f"/{digit}", exist_ok=True)

    try:
        id_url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{cid}/conformers/TXT?record_type=3d"
        response = urllib.request.urlopen(id_url)
        saveid = f"{output}/{digit}/{digit}_{lid}_{cid}_3d_id.txt"
        urllib.request.urlretrieve(id_url, saveid)
        
        response = requests.get(id_url)
        for line in response.text.splitlines():
            print(f"{line}_{cid}")
            id_cid_url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/conformers/{line}/SDF?response_type=save&response_basename=Conformer3D_COMPOUND_CID_{cid}_{line}"
            response = urllib.request.urlopen(id_cid_url)
            saveid = f"{output}/{digit}/Conformer3D_COMPOUND_{digit}_{lid}_{cid}_{line}.sdf"
            urllib.request.urlretrieve(id_cid_url, saveid)
    except:
        print('except')


0000281B00000002_10267
0000281B00000018_10267
0000281B00000019_10267
0000281B00000007_10267
0000281B00000004_10267
0000281B00000011_10267
0000281B00000001_10267
0000281B0000000B_10267
0000281B00000012_10267
0000281B00000016_10267
002D9C3E00000001_2989118
002D9C3E00000051_2989118
002D9C3E00000025_2989118
002D9C3E0000000D_2989118
002D9C3E0000005D_2989118
002D9C3E0000003D_2989118
002D9C3E00000058_2989118
002D9C3E0000000B_2989118
002D9C3E00000033_2989118
002D9C3E0000004C_2989118
0036AE8400000001_3583620
0036AE8400000021_3583620
0036AE840000002C_3583620
0036AE8400000030_3583620
0036AE8400000022_3583620
0036AE8400000020_3583620
0036AE840000000F_3583620
0036AE8400000011_3583620
0036AE8400000009_3583620
0036AE840000002D_3583620
02CF1F3A00000001_47128378
02CF1F3A00000007_47128378
02CF1F3A00000021_47128378
02CF1F3A00000018_47128378
02CF1F3A0000001D_47128378
02CF1F3A00000016_47128378
02CF1F3A00000014_47128378
02CF1F3A00000012_47128378
02CF1F3A0000001F_47128378
02CF1F3A00000011_47128378
09E8468500

# 3d 1개만 다운로드할 때
- 처음에는 isomeric smiles로 검색한 후에 없으면 cid로 검색하여 다운함.

In [7]:
# pdb, cid, lid of sdf
exist_3d_pdb = sdf['PDB'].tolist()
exist_3d_cid = sdf['CID'].tolist()
exist_3d_lid = sdf['LID'].tolist()
exist_3d_isosmi = sdf['isomeric_smiles'].tolist()

# original pdb path
ori_pdb_path = 'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb'
ori_pdb = os.listdir(ori_pdb_path)

# save pdb & 3d conformer
output = 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single'

# check = []
for digit, cid, lid, iso_smi in zip(exist_3d_pdb, exist_3d_cid, exist_3d_lid, exist_3d_isosmi):
    os.makedirs(output + f"/{digit}", exist_ok=True)
    
    pdb = digit + '.pdb'
    if pdb in ori_pdb:
        # check.append(pdb)
        sp = subprocess.Popen(f"cp -v {ori_pdb_path}/{pdb} {output}/{digit}/{digit}_{lid}_{cid}.pdb", 
                        shell=True, stdout=subprocess.PIPE,  
                        stderr=subprocess.PIPE)
    
        s, e = sp.communicate()
        print(s.decode('utf-8'))
        print(e.decode('utf-8'))
        print(sp.returncode)
    try:
        url = f'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{iso_smi}/SDF?record_type=3d&response_type=save&response_basename=Conformer3D_COMPOUND_CID_{cid}'
        response = urllib.request.urlopen(url)
        savename = f"{output}/{digit}/Conformer3D_COMPOUND_{digit}_{lid}_{cid}.sdf"
        
        # url이 가리키는 주소에 접근해서 해당 자원을 로컬 컴퓨터에 저장하기
        urllib.request.urlretrieve(url, savename)
    except HTTPError as herr:
        print(f'{digit}_{lid}_{cid} ' + 'Reason: ', herr.reason)
        url = f'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{cid}/SDF?record_type=3d&response_type=save&response_basename=Conformer3D_COMPOUND_CID_{cid}'
        response = urllib.request.urlopen(url)
        savename = f"{output}/{digit}/Conformer3D_COMPOUND_{digit}_{lid}_{cid}.sdf"
        
        # url이 가리키는 주소에 접근해서 해당 자원을 로컬 컴퓨터에 저장하기
        urllib.request.urlretrieve(url, savename)

'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FS3.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single/7FS3/7FS3_FBP_10267.pdb'


0
'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FUS.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single/7FUS/7FUS_ZIM_2989118.pdb'


0
'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FUT.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single/7FUT/7FUT_ZJ8_3583620.pdb'


0
'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FUX.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single/7FUX/7FUX_ZJR_47128378.pdb'


0
7FUX_ZJR_47128378 Reason:  PUGREST.BadRequest
'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FUY.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_output/pubchem_3d_exist/pdb_lig_single/7FUY/7FUY_ZJW_166217349.pdb'


0
'pdb_2023_human_X_ray_Prot_resol2/rcsb_api_pdb/7FV2.pdb' -> 'pdb_2023_human_X_ray_Prot_resol2/final_out

In [10]:
!jupyter nbconvert --to html preprocess_final.ipynb

[NbConvertApp] Converting notebook preprocess_final.ipynb to html
[NbConvertApp] Writing 1182002 bytes to preprocess_final.html
